### DISCUSSION TITLE
Best way to build a basic responsive main navigation (a few section with a few items) ?

### DISCUSSION THREAD

**Discussion by SebastienD11** [radix-ui/primitives#2126] [Help]
Hi ! 
Today, what would be the best way to build a basic hamburger responsive main navigation ?

The question is quite close to this one https://github.com/radix-ui/primitives/discussions/796, but I won't talk only about the Dropdown, so I created a new discussion (if that's ok...)

There is already a bunch of post about the Dropdown Menu not being really made for responsive.
You cannot make the wrapper act as an off canvas, drawer, or a full-screen modal.
It is suggested [here](https://github.com/radix-ui/primitives/issues/986) to use the Dialog component and make it behave as we want.

Ok for this part. Easy to make the dialog open and take the full screen.

My concern now if about the content of this Dialog.
If i want to create a basic vertical menu (a few section and a few link items):

![Capture d’écran, le 2023-05-02 à 15 02 46](https://user-images.githubusercontent.com/22220777/235761110-6bc5e121-2882-4c9c-8026-6ea5a5ae3fd7.png)

Looking at the available primitives for a Menu we have 3 choices :
- Dropdown Menu (uncontrolled and open by default)
- Navigation Menu
- By hand with HTML/CSS

### Dropdown Menu (uncontrolled and open by default)

This one is interesting as it comes with a lot of use case for a responsive main navigation: group, separator, label, links, keyboard up/down navigation), loop, etc.

The drawback is that even with the openDefault={true}, and by removing the portal the positioning of its content it's still relative `data-radix-popper-content-wrapper` which come with fixed position and transform style . We cannot really use this within our Dialog, right?

### Navigation Menu

At first I though this one was the right one. But here again, we have a few drawback. 
It seems to really focus on a desktop main navigation, with popover sub menu, trigger on main element, no separator, etc. 
So here as well, i can't see using this one as it's missing items (mostly the "group" concept and separator) ?

Unless maybe mixing primitives and custom html? I don't know if that's ok ? But it'd feel less complete compare to what offers the dropdown's one.

```
   <NavigationMenu.Root>
      <NavigationMenu.List>
        <NavigationMenu.Item>
          <a href="/">Home</a>
        </NavigationMenu.Item>
        <li role="separator" className="border border-grey-200" />
        <NavigationMenu.Item>
          <a href="/about">About</a>
        </NavigationMenu.Item>
      </NavigationMenu.List>
    </NavigationMenu.Root>
```

### By hand

Building a menu by hand is not that hard when it come to add all aria attribute, but with this method, we're lacking keyboard navigation and can only use tab to switch between links, unless redoing a lot of stuff that is prevent in various primitives in Radix.

I am a little lost on how to achieve what feels a basic need here ?

Thank you :)

**Comment by benoitgrelard** [comment]
Hey @SebastienD11,

I believe for our [WorkOS docs](https://workos.com/docs) on mobile we use a `NavigationMenu` inside a `Dialog`.
@vladmoroz may be able to add a bit more detail here as he implemented this I believe.

**Reply by SebastienD11** [reply]
Thanks for your insight!
I ended up using a Dropdown Menu. I was able to make it look like a responsive menu pretty easily in the end.

```
const [isOpen, setIsOpen] = useState<boolean>(false);
const onOpenChange = (open: boolean) => {
    onStateChange(open);
    setIsOpen(open);
};
  
[...]

<DropdownMenu.Root modal={true} onOpenChange={onOpenChange}>
      <DropdownMenu.Trigger asChild>
        <button
          type="button"
          className="flex grow items-center justify-center px-4"
          aria-label="Customise options"
        >
          {!isOpen && <MenuIcon width={24} height={24} />}
          {isOpen && <CloseIcon width={24} height={24} />}
        </button>
      </DropdownMenu.Trigger>
      <DropdownMenu.Portal>
        <>
          <DropdownMenu.Content
            className="relative z-30 -mt-1 max-h-[var(--responsive-menu-height)] w-screen overflow-auto bg-white"
            sideOffset={0}
            loop={true}
          >
            [...my content items]
          </DropdownMenu.Content>
          <div className="fixed inset-0 z-20 bg-black opacity-50" />
        </>
      </DropdownMenu.Portal>
    </DropdownMenu.Root>
```

The only caveat - for a classic UI with a top bar with trigger / logo / search for example on mobile, is that the Dropdown Menu popover will be relative to your trigger, the dropdown background should match this top bar for a result to be clean results. But that's the same problem with the Dialog component anyway.

Also, i added a custom overlay with the line `<div className="fixed inset-0 z-20 bg-black opacity-50" />` within my Portal in case my `DropdownMenu.Content` does not take the whole screen (minus the top bar height).

And i'm thinking `max-h-[var(--responsive-menu-height)] overflow-auto` could be replaced by the ScrollableArea primitive probably, but i didn't try it.

**Comment by vladmoroz** [comment]
As @benoitgrelard mentioned, we ourselves used a Dialog in combination with the Navigation Menu in the WorkOS docs. WorkOS docs navigation is a little complex design-wise, so we had to reuse many components between the full-size and the mobile versions, but otherwise it was a normal dialog when it comes styling.

The Dialog is used for the focus trapping, and a simple Navigation Menu just with Root, List, Item and Link parts provides the correct semantics and the arrow key support—and it doesn't have to use Trigger, Content, or Sub parts if you don't need them.
